winch/integration_test.go

```diff
@@ -0,0 +1,272 @@
+		resp.Header().Set("x-test-kedge-id", strconv.Itoa(id))
+		resp.Header().Set("x-test-auth-value", req.Header.Get("Authorization"))
+		resp.WriteHeader(http.StatusAccepted) // accepted to make sure stuff is slightly different.
+	})
+}
+
+type localKedges struct {
+	listeners []net.Listener
+	servers   []*http.Server
+}
+
+func buildAndStartServer(t *testing.T, config *tls.Config, index int) (net.Listener, *http.Server) {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err, "must be able to allocate a port for local kedge")
+	if config != nil {
+		listener = tls.NewListener(listener, config)
+	}
+	server := &http.Server{
+		// TODO(bplotka): Mimic OIDC support when added on kedge.
+		Handler: unknownPingbackHandler(index),
+	}
+	go func() {
+		server.Serve(listener)
+	}()
+	return listener, server
+}
+
+func (l *localKedges) SetupKedges(t *testing.T, config *tls.Config, num int) {
+	for i := 0; i < num; i++ {
+		listener, server := buildAndStartServer(t, config, i)
+		l.servers = append(l.servers, server)
+		l.listeners = append(l.listeners, listener)
+	}
+}
+
+func (l *localKedges) Close() error {
+	for _, l := range l.listeners {
+		l.Close()
+	}
+	return nil
+}
+
+// WinchIntegrationSuite performs tests that simulates following setup:
+// client <- plain  HTTP -> winch (forward proxy) <- TLS -> kedge (reverse proxy)
+// Kedge configuration is mimicked as OIDC authed kedge, so ClientAuth is set only to tls.VerifyClientCertIfGiven
+type WinchIntegrationSuite struct {
+	suite.Suite
+
+	winch              *http.Server
+	winchListenerPlain net.Listener
+	routes             *winch.DynamicRoutes
+
+	// Will be used to call winch.
+	winchMapper kedge_map.Mapper
+
+	localSecureKedges localKedges
+}
+
+func moveToLocalhost(addr string) string {
+	return strings.Replace(addr, "127.0.0.1", "localhost", -1)
+}
+
+func TestBackendPoolIntegrationTestSuite(t *testing.T) {
+	suite.Run(t, &WinchIntegrationSuite{})
+}
+
+func (s *WinchIntegrationSuite) SetupSuite() {
+	var err error
+	s.winchListenerPlain, err = net.Listen("tcp", "localhost:0")
+	require.NoError(s.T(), err, "must be able to allocate a port for winchListenerPlain")
+
+	s.routes = winch.NewDynamicRoutes()
+	s.winch = &http.Server{
+		Handler: winch.New(kedge_map.RouteMapper(s.routes), s.tlsClientConfigForTest()),
+	}
+	go func() {
+		s.winch.Serve(s.winchListenerPlain)
+	}()
+
+	http2ServerTlsConfig, err := connhelpers.TlsConfigWithHttp2Enabled(s.tlsServerConfigForTest())
+	if err != nil {
+		s.FailNow("cannot configure the tls config for http2")
+	}
+	// It does not make sense if kedge is not secure.
+	s.localSecureKedges.SetupKedges(s.T(), http2ServerTlsConfig, 3)
+
+	s.routes.Update(&pb.MapperConfig{
+		Routes: []*pb.Route{
+			{
+				Type: &pb.Route_Direct{
+					Direct: &pb.DirectRoute{
+						Key:      "resource1.ext.example.com",
+						KedgeUrl: "https://" + moveToLocalhost(s.localSecureKedges.listeners[0].Addr().String()),
```

Comment:
Look for the `moveToLocalhost` function - it just moved to `localhost` "domain" from `127.0.0.1` form (: